server/storage/wal/version.go

```diff
@@ -98,43 +135,106 @@ func etcdVersionFromData(entryType raftpb.EntryType, data []byte) *semver.Versio
+	return visitMessage(msg, visitor)
 }
 
-func etcdVersionFromMessage(m protoreflect.Message) *semver.Version {
-	var maxVer *semver.Version
-	md := m.Descriptor()
-	opts := md.Options().(*descriptorpb.MessageOptions)
-	if opts != nil {
-		maxVer = maxVersion(maxVer, etcdVersionFromOptionsString(opts.String()))
+func visitMessageDescriptor(md protoreflect.MessageDescriptor, visitor Visitor) error {
+	err := visitDescriptor(md, visitor)
+	if err != nil {
+		return err
+	}
+	fields := md.Fields()
+	for i := 0; i < fields.Len(); i++ {
+		fd := fields.Get(i)
+		err = visitDescriptor(fd, visitor)
+		if err != nil {
+			return err
+		}
 	}
 
+	enums := md.Enums()
+	for i := 0; i < enums.Len(); i++ {
+		err := visitEnumDescriptor(enums.Get(i), visitor)
+		if err != nil {
+			return err
+		}
+	}
+	return err
+}
+
+func visitMessage(m protoreflect.Message, visitor Visitor) error {
+	md := m.Descriptor()
+	err := visitDescriptor(md, visitor)
+	if err != nil {
+		return err
+	}
 	m.Range(func(field protoreflect.FieldDescriptor, value protoreflect.Value) bool {
 		fd := md.Fields().Get(field.Index())
-		maxVer = maxVersion(maxVer, etcdVersionFromField(fd))
+		err = visitDescriptor(fd, visitor)
+		if err != nil {
+			return false
+		}
+
 		switch m := value.Interface().(type) {
 		case protoreflect.Message:
-			maxVer = maxVersion(maxVer, etcdVersionFromMessage(m))
+			err = visitMessage(m, visitor)
 		case protoreflect.EnumNumber:
-			maxVer = maxVersion(maxVer, etcdVersionFromEnum(field.Enum(), m))
+			err = visitEnumNumber(fd.Enum(), m, visitor)
+		}
+		if err != nil {
+			return false
 		}
 		return true
 	})
-	return maxVer
+	return err
 }
 
-func etcdVersionFromEnum(enum protoreflect.EnumDescriptor, value protoreflect.EnumNumber) *semver.Version {
-	var maxVer *semver.Version
-	enumOpts := enum.Options().(*descriptorpb.EnumOptions)
-	if enumOpts != nil {
-		maxVer = maxVersion(maxVer, etcdVersionFromOptionsString(enumOpts.String()))
+func visitEnumDescriptor(enum protoreflect.EnumDescriptor, visitor Visitor) error {
+	err := visitDescriptor(enum, visitor)
+	if err != nil {
+		return err
 	}
-	valueDesc := enum.Values().Get(int(value))
-	valueOpts := valueDesc.Options().(*descriptorpb.EnumValueOptions)
+	fields := enum.Values()
+	for i := 0; i < fields.Len(); i++ {
+		fd := fields.Get(i)
+		err = visitDescriptor(fd, visitor)
+		if err != nil {
+			return err
+		}
+	}
+	return err
+}
+
+func visitEnumNumber(enum protoreflect.EnumDescriptor, number protoreflect.EnumNumber, visitor Visitor) error {
+	err := visitDescriptor(enum, visitor)
+	if err != nil {
+		return err
+	}
+	return visitEnumValue(enum.Values().Get(int(number)), visitor)
+}
+
+func visitEnumValue(enum protoreflect.EnumValueDescriptor, visitor Visitor) error {
+	valueOpts := enum.Options().(*descriptorpb.EnumValueOptions)
 	if valueOpts != nil {
-		maxVer = maxVersion(maxVer, etcdVersionFromOptionsString(valueOpts.String()))
+		ver, _ := etcdVersionFromOptionsString(valueOpts.String())
```

Comment:
I wonder whether we are not parsing these options too frequently (e.g. on each interpreted _data_ enum). 
Alternatively we could build a map(s):
  -  field-> version
  - enum -> version 
The way we build the version file... and during the data processing just keep performing lookups in the map.

Reply:
Sure, we will definitely need this for version API calls as they will be latency sensitive, however I would skip this optimization of static analysis.

Reply:
Aren't we afraid that  start time of etcd server that will keep processing WAL log (ReadWALVersion) will get significantly slower. Have you tested it with realistic: O(64MB) Wal files ?

Reply:
We don't verify WAL during server start, we do it during downgrade and after we created a snapshot.